Remove gemfile_entry from GeneratorAppBase:

roramirez opened the pull request:
This private method is not in used. It's introduced in 3f2230b but at the
moment there not sense to keeping this code.


Ref: https://github.com/rails/rails/issues/36699#issuecomment-596806887

roramirez committed:
Remove gemfile_entry from GeneratorAppBase:

This private method is not in used. It's introduced in 3f2230b but at the
moment there not sense to keeping this code.

rafaelfranca commented:
Thank you for the pull request but this method is to be used in application templates, not by the framework, so it still makes sense to exist.

roramirez commented:
Hi @rafaelfranca 

I've trying to figure out how get these lines of code as mentioned as application template but I don't get it. https://guides.rubyonrails.org/rails_application_templates.html 

Can you provider a example how I can use this code by application template? I'll could prepare a test to covered these lines.

rafaelfranca commented:
Adding a method call like `gemfile_entry('rack')` in the template.rb is how it should work.

roramirez commented:
Mm.... really doesn't work. The  `gemfile_entry` is call after the render the `Gemfile.tt`

Maybe there a way to inject this before to called the Gemfile.tt but as the way to how should add extra gem by this method not worked as suspected how should to be.


Log here

**Diff** 
```diff
diff --git a/railties/lib/rails/generators/app_base.rb b/railties/lib/rails/generators/app_base.rb
index 29082bcce7..297c37d824 100644
--- a/railties/lib/rails/generators/app_base.rb
+++ b/railties/lib/rails/generators/app_base.rb
@@ -124,10 +124,12 @@ def gemfile_entry(name, *args) # :doc:
         else
           @extra_entries << GemfileEntry.version(name, version)
         end
+        puts "Extras HERE #{@extra_entries}"
         self
       end

       def gemfile_entries # :doc:
+        puts "Extras in gemfile_entries :#{@extra_entries}"
         [rails_gemfile_entry,
          database_gemfile_entry,
          web_server_gemfile_entry,

```

**Template file**
```zsh
vagrant@railsdev:~/rails$ cat /tmp/template.rb 
gem_group :test do
    gem "launchy"
end

gemfile_entry('rack')
```

**New application using template**
```zsh
vagrant@railsdev:~/rails$ bundle exec rails new  --quiet --template=/tmp/template.rb  /tmp/rails/gemfile_entry
Extras in gemfile_entries :[]
Extras in gemfile_entries :[]
Extras HERE [#<struct Rails::Generators::AppBase::GemfileEntry name="rack", version=nil, comment=nil, options={}, commented_out=false>]
The dependency tzinfo-data (>= 0) will be unused by any of the platforms Bundler is installing for. Bundler is installing for ruby but the dependency is only for x86-mingw32, x86-mswin32, x64-mingw32, java. To add those platforms to the bundle, run `bundle lock --add-platform x86-mingw32 x86-mswin32 x64-mingw32 java`.
Could not find gem 'rails (~> 6.1.0.alpha)' in any of the gem sources listed in your Gemfile.
Could not find gem 'rails (~> 6.1.0.alpha)' in any of the gem sources listed in your Gemfile.
bundler: command not found: spring
Install missing gem executables with `bundle install`
Webpacker requires Node.js >= 8.16.0 and you are using 5.12.0
Please upgrade Node.js https://nodejs.org/en/download/


```
**Gemfile output**
```zsh
vagrant@railsdev:~/rails$ cat /tmp/rails/gemfile_entry/Gemfile 
source 'https://rubygems.org'
git_source(:github) { |repo| "https://github.com/#{repo}.git" }

ruby '2.6.5'

# Bundle edge Rails instead: gem 'rails', github: 'rails/rails'
gem 'rails', '~> 6.1.0.alpha'
# Use sqlite3 as the database for Active Record
gem 'sqlite3', '~> 1.4'
# Use Puma as the app server
gem 'puma', '~> 4.1'
# Use SCSS for stylesheets
gem 'sass-rails', '>= 6'
# Transpile app-like JavaScript. Read more: https://github.com/rails/webpacker
gem 'webpacker', '~> 4.0'
# Turbolinks makes navigating your web application faster. Read more: https://github.com/turbolinks/turbolinks
gem 'turbolinks', '~> 5'
# Build JSON APIs with ease. Read more: https://github.com/rails/jbuilder
gem 'jbuilder', '~> 2.7'
# Use Redis adapter to run Action Cable in production
# gem 'redis', '~> 4.0'
# Use Active Model has_secure_password
# gem 'bcrypt', '~> 3.1.7'

# Use Active Storage variant
# gem 'image_processing', '~> 1.2'

# Reduces boot times through caching; required in config/boot.rb
gem 'bootsnap', '>= 1.4.4', require: false

group :development, :test do
  # Call 'byebug' anywhere in the code to stop execution and get a debugger console
  gem 'byebug', platforms: [:mri, :mingw, :x64_mingw]
end

group :development do
  # Access an interactive console on exception pages or by calling 'console' anywhere in the code.
  gem 'web-console', '>= 3.3.0'
  # Display performance information such as SQL time and flame graphs for each request in your browser. Can be configured to work on production as well see: https://github.com/MiniProfiler/rack-mini-profiler/blob/master/README.md
  gem 'rack-mini-profiler', '~> 2.0.0'
  gem 'listen', '~> 3.2'
  # Spring speeds up development by keeping your application running in the background. Read more: https://github.com/rails/spring
  gem 'spring'
end

group :test do
  # Adds support for Capybara system testing and selenium driver
  gem 'capybara', '>= 3.26'
  gem 'selenium-webdriver'
  # Easy installation and use of web drivers to run system tests with browsers
  gem 'webdrivers'
end

# Windows does not include zoneinfo files, so bundle the tzinfo-data gem
gem 'tzinfo-data', platforms: [:mingw, :mswin, :x64_mingw, :jruby]

group :test do
  gem 'launchy'
end
```

rafaelfranca commented:
In that case `add_gem_entry_filter` is probably also not working. https://github.com/rails/rails/commit/f7e0758714e645865361b5566253d3e7d896beff

@tenderlove do you remember how those features should be used?

roramirez commented:
Interesting. I'm gone to the past at Rails 4.1.0 where the feature are introduced and I've the same result. Maybe was a going a be a feature to be use but at this moment isn't finished or I'm wrong how to use

```zsh
vagrant@railsdev:~$ rails -v
Deprecation warning: Expected string default value for '--rc'; got false (boolean).
This will be rejected in the future unless you explicitly pass the options `check_default_type: false` or call `allow_incompatible_default_type!` in your code
You can silence deprecations warning by setting the environment variable THOR_SILENCE_DEPRECATION.
Rails 4.1.0
vagrant@railsdev:~$ cat /tmp/template.rb 
gem_group :test do
    gem "launchy"
end

gemfile_entry('rack')

vagrant@railsdev:~$ rails new  --quiet --template=/tmp/template.rb  /tmp/rails/gemfile_entry_4.1.0
Deprecation warning: Expected string default value for '--rc'; got false (boolean).
This will be rejected in the future unless you explicitly pass the options `check_default_type: false` or call `allow_incompatible_default_type!` in your code
You can silence deprecations warning by setting the environment variable THOR_SILENCE_DEPRECATION.
vagrant@railsdev:~$ cat /tmp/rails/gemfile_entry
gemfile_entry/       gemfile_entry_4.1.0/ 
vagrant@railsdev:~$ cat /tmp/rails/gemfile_entry
gemfile_entry/       gemfile_entry_4.1.0/ 
vagrant@railsdev:~$ cat /tmp/rails/gemfile_entry_4.1.0/Gemfile 
source 'https://rubygems.org'


# Bundle edge Rails instead: gem 'rails', github: 'rails/rails'
gem 'rails', '4.1.0'
# Use sqlite3 as the database for Active Record
gem 'sqlite3'
# Use SCSS for stylesheets
gem 'sass-rails', '~> 4.0.3'
# Use Uglifier as compressor for JavaScript assets
gem 'uglifier', '>= 1.3.0'
# Use CoffeeScript for .js.coffee assets and views
gem 'coffee-rails', '~> 4.0.0'
# See https://github.com/sstephenson/execjs#readme for more supported runtimes
# gem 'therubyracer',  platforms: :ruby

# Use jquery as the JavaScript library
gem 'jquery-rails'
# Turbolinks makes following links in your web application faster. Read more: https://github.com/rails/turbolinks
gem 'turbolinks'
# Build JSON APIs with ease. Read more: https://github.com/rails/jbuilder
gem 'jbuilder', '~> 2.0'
# bundle exec rake doc:rails generates the API under doc/api.
gem 'sdoc', '~> 0.4.0',          group: :doc

# Spring speeds up development by keeping your application running in the background. Read more: https://github.com/rails/spring
gem 'spring',        group: :development

# Use ActiveModel has_secure_password
# gem 'bcrypt', '~> 3.1.7'

# Use unicorn as the app server
# gem 'unicorn'

# Use Capistrano for deployment
# gem 'capistrano-rails', group: :development

# Use debugger
# gem 'debugger', group: [:development, :test]


group :test do
  gem "launchy"
end
```

roramirez commented:
> In that case `add_gem_entry_filter` is probably also not working. [f7e0758](https://github.com/rails/rails/commit/f7e0758714e645865361b5566253d3e7d896beff)

I tried in the template and how you are mentioned here https://github.com/rails/rails/pull/15716#issuecomment-46308751 and neither

roramirez commented:
I'll want to keep open this PR. cc @rafaelfranca

rafaelfranca commented:
Let's also remove `add_gem_entry_filter` and related code in this PR.

roramirez committed:
Remove add_gem_entry_filter from GeneratorAppBase:

The add_gem_entry_filter was introduce in
f7e0758714e645865361b5566253d3e7d896beff and doesn't work this
feature

roramirez commented:
@rafaelfranca I've just remove this code in separate commit.
